### PR TITLE
Initial implementation of mm-experiments

### PR DESCRIPTION
**Pull request by TiddoLangerak**
This is the initial implementation for mm-experiments. It contains a generalized implementation for experiments, as well as a specific Google Experiments adapter to load experiments from Google Analytics.

@Magnetme/monolith - RFR!

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,4 +1,157 @@
...
+		};
+	})
+	.factory('googleExperiments', ($q, $rootScope) => {
+		'ngInject';
+		const experiments = new Map();
+		const resolvers = new Map();
+		const promises = new Map();
+
+		window.mmGoogleExperimentCallback = (id, variation) => {
+			$rootScope.$apply(() => {
```
Comment (alexnederlof) [inline src/index.js]: why not regular `$scope.apply`?

Comment (TiddoLangerak) [inline src/index.js]: We're in a factory here, which doesn't have it's own `$scope`.

File: src/index.js [hunk 2]
```diff
@@ -1,4 +1,157 @@
 import angular from 'angular';
 
-export default angular.module("mm.mm-experiments", []);
-	
+export default angular.module("mm.experiments", [])
+	.factory('experiments', ($q) => {
```
Comment (wouter-willems) [inline src/index.js]: you can leave out the `()`

File: src/index.js [hunk 3]
```diff
@@ -1,4 +1,157 @@
...
+			}
+		};
+	})
+	.factory('googleExperiments', ($q, $rootScope) => {
+		'ngInject';
+		const experiments = new Map();
+		const resolvers = new Map();
+		const promises = new Map();
+
+		window.mmGoogleExperimentCallback = (id, variation) => {
```
Comment (wouter-willems) [inline src/index.js]: why does this need to be on the window object?

**Comment by alexnederlof** [comment]
:+1:

File: src/index.js [hunk 4]
```diff
@@ -1,4 +1,157 @@
...
+			}
+		}
+	})
+	.directive('mmVariation', () => {
+		return {
+			restrict : 'A',
+			require : '^mmExperiment',
+			transclude : 'element',
+			priority: 599, //1 less than ng-if
+			terminal : true,
```
Comment (wouter-willems) [inline src/index.js]: why is this needed?

File: src/index.js [hunk 5]
```diff
@@ -1,4 +1,157 @@
...
+				ctrl._setExperiment(attrs.mmExperiment);
+				scope.$variation = await experiments.getVariation(attrs.mmExperiment);
+			}
+		}
+	})
+	.directive('mmVariation', () => {
+		return {
+			restrict : 'A',
+			require : '^mmExperiment',
+			transclude : 'element',
```
Comment (wouter-willems) [inline src/index.js]: correct my if i am wrong, but transclude is a boolean property, right

**Comment by wouter-willems** [comment]
I find it hard to review cause I am missing context. How do these directives work? can you provide examples?

File: src/index.js (hunk 3 above)
Comment (TiddoLangerak) [inline src/index.js]: Because it needs to be called from the iframe. Can't call angular functions from there.

File: src/index.js (hunk 4 above)
Comment (TiddoLangerak) [inline src/index.js]: Because adding this directive to an element will transclude the entire directive, not just it's content, and we want to evaluate all directives on the element only when they are transcluded. I.e. suppose we have 

```html
<a mm-variation='0' mm-image-zoom>
    <img src="image-one.jpg" />
</a>
```
Then we don't want `mm-image-zoom` to be executed if the variation is not `0`. By adding `terminal : true` we ensure that it will not be executed on load, but only when we explicitly transclude it. 

I've stolen this from `ng-if` by the way, my intention was to make this work as similar as possible as `ng-if`.

File: src/index.js (hunk 5 above)
Comment (TiddoLangerak) [inline src/index.js]: `transclude : true` transcludes it's content, `transclude : 'element'` transcludes the entire element.

**Comment by TiddoLangerak** [comment]
Examples are in the [README](https://github.com/Magnetme/experiments/pull/1/files#diff-04c6e90faac2675aa89e2176d2eec7d8R11).

File: src/index.js (hunk 2 above)
Comment (TiddoLangerak) [inline src/index.js]: I know, but I personally prefer to leave them in in this particular case because I don't really consider the factory function to be a single argument function. But if you want to I can remove them.

File: src/index.js (hunk 3 above)
Comment (wouter-willems) [inline src/index.js]: haha, that makes sense

File: src/index.js (hunk 2 above)
Comment (wouter-willems) [inline src/index.js]: Nah, leave it

File: src/index.js (hunk 4 above)
Comment (wouter-willems) [inline src/index.js]: I am not sure what transclusion has to do with this. As far as I know, `terminal : true` will simply ignore all other directives on the same element with a lower prio

File: src/index.js (hunk 5 above)
Comment (wouter-willems) [inline src/index.js]: ah, did not know this. Looked a bit further into it, looks like it kinda works like a 'replace'

**Comment by wouter-willems** [comment]
Codewise it is :+1: 
But I must say that I don't fully understand the api. The example does not really tell me when to use deferred and when to use a normal variation

File: src/index.js (hunk 4 above)
Comment (TiddoLangerak) [inline src/index.js]: Exactly, but due to transclusion they do get applied at a later state. Without transclusion they would just be eliminated altogether, which we obviously don't want.

**Comment by TiddoLangerak** [comment]
Deferred is for when the variation comes from a promise. Will clarify it a bit.
